src/platform/ESP32/MdnsImpl.cpp

```diff
@@ -0,0 +1,241 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "lib/protocols/mdns/platform/Mdns.h"
+
+#include <esp_err.h>
+#include <lwip/ip4_addr.h>
+#include <lwip/ip6_addr.h>
+#include <mdns.h>
+
+#include "platform/CHIPDeviceLayer.h"
+#include "support/CHIPMem.h"
+#include "support/CodeUtils.h"
+#include "support/logging/CHIPLogging.h"
+
+namespace {
+
+static constexpr uint32_t kTimeoutMilli = 3000;
+static constexpr size_t kMaxResults     = 20;
+
+} // namespace
+
+namespace chip {
+namespace Protocols {
+namespace Mdns {
+
+CHIP_ERROR ChipMdnsInit(MdnsAsnycReturnCallback initCallback, MdnsAsnycReturnCallback errorCallback, void * context)
+{
+    CHIP_ERROR error = CHIP_NO_ERROR;
+    esp_err_t espError;
+    uint16_t discriminator;
+    char hostname[11]; // Hostname willbe "CHIP-XXXX"
+
+    espError = mdns_init();
+    VerifyOrExit(espError == ESP_OK, error = CHIP_ERROR_INTERNAL);
+    SuccessOrExit(error = DeviceLayer::ConfigurationMgr().GetSetupDiscriminator(discriminator));
+    snprintf(hostname, sizeof(hostname), "CHIP-%04d", discriminator);
+    espError = mdns_hostname_set(hostname);
+    VerifyOrExit(espError == ESP_OK, error = CHIP_ERROR_INTERNAL);
+
+exit:
+    if (espError != ESP_OK)
+    {
+        ChipLogError(DeviceLayer, "esp mdns internal error: %s", esp_err_to_name(espError));
+    }
+    initCallback(context, error);
+
+    return error;
+}
+
+static const char * GetProtocolString(MdnsServiceProtocol protocol)
+{
+    return protocol == MdnsServiceProtocol::kMdnsProtocolTcp ? "_tcp" : "_udp";
+}
+
+CHIP_ERROR ChipMdnsPublishService(const MdnsService * service)
+{
+    CHIP_ERROR error        = CHIP_NO_ERROR;
+    mdns_txt_item_t * items = nullptr;
+    esp_err_t espError;
+
+    VerifyOrExit(service->mTextEntrySize <= UINT8_MAX, error = CHIP_ERROR_INVALID_ARGUMENT);
+    if (service->mTextEntryies)
+    {
+        items = static_cast<mdns_txt_item_t *>(chip::Platform::MemoryCalloc(service->mTextEntrySize, sizeof(mdns_txt_item_t)));
+        VerifyOrExit(items != nullptr, error = CHIP_ERROR_NO_MEMORY);
+        for (size_t i = 0; i < service->mTextEntrySize; i++)
+        {
+            items[i].key = service->mTextEntryies[i].mKey;
+            // Unfortunately ESP mdns stack dosen't support arbitrary binary data
+            items[i].value = reinterpret_cast<const char *>(service->mTextEntryies[i].mData);
```

Comment:
looks like another dangling pointer.  in the only call to this function in this PR, the mData passed in is a pointer to a value on the stack.

this API needs the memory responsibilities spelled out.  strdup() for these would be ok, but I think it'd be better to perhaps use a Cancelable, where the registrar maintains the memory (via RAII), but can cancel the registration with a simple destructor

Reply:
actually, looks like these datastructures exist solely to serve as a type impedance match between the CHIP Mdns API and the esp32 API?

Reply:
As in the previous comment, it's safe to pass a stack pointer here and have esp library to do the strdup for us.

Reply:
yeap, got it.  can you stack-allocate the items array instead of going to the heap?